contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,758 @@
+  {
+    uint256 startGas = gasleft();
+    (
+      bytes32 keyHash,
+      uint256 requestId,
+      uint256 randomness,
+      FulfillmentParams memory fp
+    ) = getRandomnessFromProof(proof);
+
+
+    uint256[] memory randomWords = new uint256[](fp.numWords);
+    for (uint256 i = 0; i < fp.numWords; i++) {
+      randomWords[i] = uint256(keccak256(abi.encode(randomness, i)));
+    }
+
+    // Prevent re-entrancy. The user callback cannot call fulfillRandomWords again
+    // with the same proof because this getRandomnessFromProof will revert because the requestId
+    // is gone.
+    delete s_commitments[requestId];
+    VRFConsumerBaseV2 v;
+    bytes memory resp = abi.encodeWithSelector(v.rawFulfillRandomWords.selector, requestId, randomWords);
+    uint256 gasPreCallback = gasleft();
+    if (gasPreCallback < fp.callbackGasLimit) {
+      revert InsufficientGasForConsumer(gasPreCallback, fp.callbackGasLimit);
+    }
+    // Call with explicitly the amount of callback gas requested
+    // Important to not let them exhaust the gas budget and avoid oracle payment.
+    bool success = callWithExactGas(fp.callbackGasLimit, fp.sender, resp);
+    emit RandomWordsFulfilled(requestId, randomWords, success);
+
+    // We want to charge users exactly for how much gas they use in their callback.
+    // The gasAfterPaymentCalculation is meant to cover these additional operations where we
+    // decrement the subscription balance and increment the oracles withdrawable balance.
+    // We also add the flat link fee to the payment amount.
+    // Its specified in millionths of link, if s_config.fulfillmentFlatFeeLinkPPM = 1
+    // 1 link / 1e6 = 1e18 juels / 1e6 = 1e12 juels.
+    uint96 payment = calculatePaymentAmount(startGas, s_config.gasAfterPaymentCalculation, s_config.fulfillmentFlatFeeLinkPPM, tx.gasprice);
+    if (s_subscriptions[fp.subId].balance < payment) {
+      revert InsufficientBalance();
+    }
+    s_subscriptions[fp.subId].balance -= payment;
+    s_withdrawableTokens[s_serviceAgreements[keyHash]] += payment;
+  }
+
+  // Get the amount of gas used for fulfillment
+  function calculatePaymentAmount(
+      uint256 startGas,
+      uint256 gasAfterPaymentCalculation,
+      uint32  fulfillmentFlatFeeLinkPPM,
+      uint256 weiPerUnitGas
+  )
+    private
+    view
+    returns (
+      uint96
+    )
+  {
+    int256 weiPerUnitLink;
+    weiPerUnitLink = getFeedData();
+    if (weiPerUnitLink < 0) {
+      revert InvalidFeedResponse(weiPerUnitLink);
+    }
+    // (1e18 juels/link) (wei/gas * gas) / (wei/link) = jules
+    uint256 paymentNoFee = 1e18*weiPerUnitGas*(gasAfterPaymentCalculation + startGas - gasleft()) / uint256(weiPerUnitLink);
+    uint256 fee = 1e12*uint256(fulfillmentFlatFeeLinkPPM);
+    if (paymentNoFee > (1e27-fee)) {
+      revert PaymentTooLarge(); // Payment + fee cannot be more than all of the link in existence.
+    }
+    return uint96(paymentNoFee+fee);
+  }
+
+  function getRandomnessFromProof(
```

Comment:
Test helper could help unit test this also, but if you feel it's adequately covered in the go tests, then that's your call.

Reply:
Yeah testing this in typescript runs up against the same problem as testing the whole fulfillRandomWords. If you don't port the proof generating code, then you need to use precalculated proofs but the proof depends on the blockhash but you can't fix the blockhash at a specific value. There are ways to test parts of it if you restructure the code to keep the blockhash part of out if OR if you force it to use the blockhash store by mining 256 blocks in the test. I don't think those approaches are worth the trouble given there is extensive v1 go tests for the actual proof->random value part

Reply:
Added some fulfillRandomWords unhappy path tests